histomicstk/features/FeatureExtraction.py

```diff
@@ -73,374 +78,104 @@ def FeatureExtraction(Label, In, Ic, K=128, Fs=6, Delta=8):
-        )
-        # calculate and add FSDs
-        FSDGroup[regionIdx, :] = FSDs(
-            Bounds[:, 0], Bounds[:, 1],
-            K, Interval
-        )
+            Label[min_row:max_row, min_col:max_col] == regions[i].label
+        ).astype(np.bool)
+        # compute Fourier shape descriptors
+        FSDGroup[i, :] = ComputeFSDs(Nucleus, K, Fs)
         # generate object coords for nuclei and cytoplasmic regions
-        Nuclei[regionIdx] = region.coords
+        Nuclei = regions[i].coords
+        # compute Texture, Gradient, Intensity features
+        HematoxylinGradientGroup[i, :] = \
+            ComputeGradientFeatures(In, Nuclei, diffGn, BW_cannyn)
+        HematoxylinIntensityGroup[i, :] = \
+            ComputeIntensityFeatures(In, Nuclei)
         # get mask for all nuclei in neighborhood
         Mask = (
             Label[min_row:max_row, min_col:max_col] > 0
         ).astype(np.uint8)
         # remove nucleus region from cytoplasm+nucleus mask
         cytoplasm = (
-            np.logical_xor(Mask, dilation(Nucleus, Disk))
-        ).astype(np.uint8)
+            np.logical_xor(Mask, binary_dilation(Nucleus, Disk))
+        )
         # get list of cytoplasm pixels
-        Cytoplasms[regionIdx] = GetPixCoords(
-            cytoplasm, min_row, min_col)
-        regionIdx = regionIdx + 1
-
-    # calculate hematoxlyin features, capture feature names
-    HematoxylinIntensityGroup = IntensityFeatureGroup(In, Nuclei)
-    HematoxylinTextureGroup = TextureFeatureGroup(In, Nuclei)
-    HematoxylinGradientGroup = GradientFeatureGroup(In, Nuclei)
-    # calculate eosin features
-    EosinIntensityGroup = IntensityFeatureGroup(Ic, Cytoplasms)
-    EosinTextureGroup = TextureFeatureGroup(Ic, Cytoplasms)
-    EosinGradientGroup = GradientFeatureGroup(Ic, Cytoplasms)
-
-    # add columns to dataframe
-    df['X'] = CentroidX
-    df['Y'] = CentroidY
-
-    df['Area'] = Area
-    df['Perimeter'] = Perimeter
-    df['Eccentricity'] = Eccentricity
-    df['Circularity'] = Circularity
-    df['MajorAxisLength'] = MajorAxisLength
-    df['MinorAxisLength'] = MinorAxisLength
-    df['Extent'] = Extent
-    df['Solidity'] = Solidity
+        regionCoords = np.argwhere(cytoplasm == 1)
+        regionCoords[:, 0] = regionCoords[:, 0] + min_row
+        regionCoords[:, 1] = regionCoords[:, 1] + min_col
+        # compute Texture, Gradient, Intensity features
+        EosinGradientGroup[i, :] = \
+            ComputeGradientFeatures(Ic, regionCoords, diffGc, BW_cannyc)
+        EosinIntensityGroup[i, :] = \
+            ComputeIntensityFeatures(Ic, regionCoords)
 
-    for i in range(0, Fs):
-        df['FSD' + str(i+1)] = FSDGroup[:, i]
+    # initialize panda dataframe
+    df = pd.DataFrame()
 
-    for f in HematoxylinIntensityGroup._fields:
-        df['Hematoxylin' + f] = getattr(HematoxylinIntensityGroup, f)
+    fmorph = ComputeMorphometryFeatures(Label)
+    df = pd.concat([df, fmorph], axis=1)
```

Comment:
We can join the dataframes returned by all the feature group functions using pandas.concat at once as shown here.